Add missing data lost during import from Whitehall to Content Publisher

brucebolt opened the pull request:
We were missing some data when importing a document from Whitehall to Content Publisher.

This PR adds the following:
- Imported editions are marked with a `backdated_to` timestamp where appropriate (when the edition was backdated in Whitehall, determined by comparing the first published timestamp with the time of the first publish event)
- Import the `first_published_at` date for the document (taken from the first publish event on the first edition of a document; cannot be overridden by backdating)
- Editions have a `last_edited_at` timestamp (most recent edit event from the Whitehall `revision_history`)
- Imported editions are not marked as `revision_synced` (i.e. not marked as synced with Publishing API, which has not been completed at the import stage)

Trello card: https://trello.com/c/1RIiRBUF

kevindew commented on spec/lib/whitehall_importer/create_edition_spec.rb:
```diff
@@ -173,6 +180,7 @@
 
         expect(edition.created_at).to eq(created_at)
         expect(edition.updated_at).to eq(updated_at)
+        expect(edition.last_edited_at).to eq(whitehall_edition.dig("revision_history", -1, "created_at"))
```
This dig is a bit cryptic. I think we're expecting it to be the same as the updated_at value that we set it in a let? might be a bit cleaner to compare to that.

brucebolt commented on spec/lib/whitehall_importer/create_revision_spec.rb:
```diff
@@ -31,6 +31,23 @@
+
+      revision = described_class.call(document, whitehall_edition)
+
+      expect(revision.backdated_to).to eq("2015-10-03 12:34:00 +0000")
```
I've updated this test to use an actual timestamp

brucebolt commented:
I've removed the remaining `.fetch(foo, nil)` and updated a line of code to use the `EditionHistory` class.

We can review whether a `state_events` method would be more suitable if/when we find a need for a non-first or last event.

kevindew commented on lib/whitehall_importer/import.rb:
```diff
@@ -61,8 +61,22 @@ def create_document(user_ids)
+
+    def history(edition)
+      @history ||= EditionHistory.new(edition["revision_history"])
+    end
```
You can't use a memo with an argument as it'll be ignored on subsequent calls.

Since you're only accessing first_edition for history you might just want to have a first_edition_history method and drop the first_edition one?

brucebolt replied:
Good spot!  I've updated to have a `first_edition_history` method instead

kevindew commented on lib/whitehall_importer/import.rb:
```diff
@@ -52,7 +52,7 @@ def create_document(user_ids)
       end
 
-      event = whitehall_document["editions"].first["revision_history"].select { |h| h["event"] == "create" }.first
+      event = first_edition_history.create_event
```
Argh sorry to be a pain - I didn't notice this before. But if we use the bang equivalent here we can drop the need for the raise below this.

brucebolt replied:
I've pushed a change to reflect this
